File: Training/Advanced_Cluster_Usage/Exercise3/Fortran/omp_pi.f90
!+++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++
! Program: omp_pi.f90
!          OpenMP implementation of Monte-Carlo algorithm for calculating PI
!          Translated from omp_pi.c, adapted for gfortran
!
! Compile: gfortran -o omp_pi.x omp_pi.f90 -O2 -fopenmp
! 
! Run:     ./omp_pi.x <number_of_samples> <number_of_threads>
!          OMP_NUM_THREADS=<number_of_threads> ./omp_pi.x (optional override)
!+++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++
PROGRAM omp_pi
  USE OMP_LIB   ! OpenMP library
  IMPLICIT NONE
  
  INTEGER(4) :: i, count, samples, nthreads, tid
  INTEGER(4) :: seed_size
  INTEGER(4), ALLOCATABLE :: seed(:)
  REAL(8)    :: x, y, z
  REAL(8)    :: t0, t1, tf, PI
  REAL(8), PARAMETER :: PI_EXACT = 3.14159265358979323846D0
  CHARACTER(LEN=20) :: arg1, arg2
  
  ! Get command-line arguments
  CALL GET_COMMAND_ARGUMENT(1, arg1)
  CALL GET_COMMAND_ARGUMENT(2, arg2)
  READ(arg1, *) samples   ! Number of samples
  READ(arg2, *) nthreads  ! Number of threads
  
  CALL OMP_SET_NUM_THREADS(nthreads)
  
  WRITE(*,'(A,I2)') "Number of threads: ", nthreads
  
  t0 = OMP_GET_WTIME()
  count = 0
  
  !$OMP PARALLEL PRIVATE(i, x, y, z, tid, seed) SHARED(samples)
    tid = OMP_GET_THREAD_NUM()
    
    ! Thread-specific seed for intrinsic RNG
    CALL RANDOM_SEED(SIZE=seed_size)
    ALLOCATE(seed(seed_size))
    seed = 1202107158 + tid * 1999  ! Match C seeding strategy
    CALL RANDOM_SEED(PUT=seed)
    DEALLOCATE(seed)
    
    !$OMP DO REDUCTION(+:count)
    DO i = 0, samples - 1           ! Match C loop range (0 to samples-1)
      CALL RANDOM_NUMBER(x)
      CALL RANDOM_NUMBER(y)
      z = x*x + y*y
      IF (z <= 1.0D0) count = count + 1
    END DO
    !$OMP END DO
  !$OMP END PARALLEL
  
  t1 = OMP_GET_WTIME()
  tf = t1 - t0
  
  ! Estimate PI
  PI = 4.0D0 * REAL(count, KIND=8) / REAL(samples, KIND=8)
  
  WRITE(*,'(A,F7.5)') "Exact value of PI: ", PI_EXACT
  WRITE(*,'(A,F7.5)') "Estimate of PI:    ", PI
  WRITE(*,'(A,F7.2,A)') "Time: ", tf, " sec."
  
END PROGRAM omp_pi

File: Training/Advanced_Cluster_Usage/Exercise3/Cpp/omp_pi.cpp
#include <iostream>
#include <iomanip>    // Added for setprecision
#include <random>
#include <chrono>
#include <cmath>
#include <omp.h>

int main(int argc, char *argv[]) {
    if (argc != 3) {
        std::cerr << "Usage: " << argv[0] << " <number_of_samples> <number_of_threads>" << std::endl;
        return 1;
    }

    int samples = std::atoi(argv[1]);    // Number of samples
    int nthreads = std::atoi(argv[2]);   // Number of threads
    omp_set_num_threads(nthreads);

    std::cout << "Number of threads: " << nthreads << std::endl;

    // Get start time
    auto t0 = omp_get_wtime();
    int count = 0;

#pragma omp parallel
    {
        // Each thread gets its own random number generator
        unsigned int seed = 1202107158 + omp_get_thread_num() * 1999;
        std::mt19937 gen(seed);  // Mersenne Twister random number generator
        std::uniform_real_distribution<double> dist(0.0, 1.0);

#pragma omp for reduction(+:count)
        for (int i = 0; i < samples; i++) {
            double x = dist(gen);
            double y = dist(gen);
            double z = x * x + y * y;
            if (z <= 1.0) {
                count++;
            }
        }
    }

    // Get end time
    auto t1 = omp_get_wtime();
    double tf = t1 - t0;

    // Estimate PI
    double PI = 4.0 * count / samples;

    // Output results
    std::cout << std::fixed << std::setprecision(5);
    std::cout << "Exact value of PI: " << M_PI << std::endl;
    std::cout << "Estimate of PI:    " << PI << std::endl;
    std::cout << std::fixed << std::setprecision(2);
    std::cout << "Time: " << tf << " sec." << std::endl;

    return 0;
}
